### DISCUSSION TITLE
How to differentiate between SQL statement types in `TypeDecorator.process_bind_param`?

### DISCUSSION THREAD

**Discussion by maximvlah** [sqlalchemy/sqlalchemy#11697] [Usage Questions]
We implemented a custom type:
```python

from sqlalchemy import Dialect, types
from typing import Any
from email_validator import (
    EmailNotValidError,
    EmailSyntaxError,
    EmailUndeliverableError,
    validate_email,
)

class EmailValidationError(Exception):
    pass

class Email(str):
    def __new__(cls, value: Any) -> Email:
        try:
            validate_email(str(value))
        except EmailSyntaxError as e:
            raise EmailValidationError("Invalid email")
        except EmailUndeliverableError as e:
            raise EmailValidationError(
                "Invalid email domain, check the part after the @ symbol."
            )
        except EmailNotValidError as e:
            raise EmailValidationError("Invalid email")

        return super().__new__(cls, str(value))

class EmailType(types.TypeDecorator):
    impl = types.Unicode
    cache_ok = True

    @property
    def python_type(self) -> type[Email]:
        return Email

    def process_bind_param(self, value: Any, dialect: Dialect) -> str | None:
        if value is None:
            return

        elif isinstance(value, Email):
            return value.lower()

        return Email(value).lower()

    def process_result_value(self, value: Any, dialect: Dialect) -> Email | None:
        if value and isinstance(value, str):
            return Email(value)
        return None

```

According to the `process_bind_param` [docs](https://docs.sqlalchemy.org/en/20/core/custom_types.html#sqlalchemy.types.TypeDecorator.process_bind_param), we can use this method for validating data:
_The operation could be anything desired to perform custom behavior, such as transforming or serializing data. **This could also be used as a hook for validating logic.**_

This is exactly what we are doing with our custom `Email` type and everything works fine when we `write/read` the  database records with the fields of type `Email`. 

However, when we are running search statements, i.e. sth like 

`select(User).where(User.email.ilike("%test@mail.com%"))`

the email value [has to be wrapped with `%` signs](https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.ColumnOperators.ilike) which then when passed through `return Email(value).lower()` in `process_bind_param`, raises the `EmailValidationError`.

So we would like to be able to differentiate between the search and write statements and handle them differently within the `process_bind_param`. 

Is there a proper sqlalchemy way to do so?

Thanks!

**Comment by CaselIT** [comment]
Hi,

I don't think you can solve this using process_bind_param, since it does not what bind param is coercing.
What you can do is customize the logic using other methods, like https://docs.sqlalchemy.org/en/13/core/custom_types.html#sqlalchemy.types.TypeDecorator.coerce_compared_value
